magicgui/_magicgui.py

```diff
@@ -144,36 +187,176 @@ def magicgui(
     >>> my_function.a.value == 1  # True
     >>> my_function.b.value = 'world'
     """
-    if "result" in param_options:
-        warn(
-            "\n\nThe 'result' option is deprecated and will be removed in the future."
-            "Please use `result_widget=True` instead.\n",
-            FutureWarning,
-        )
+    return _magicgui(**locals())
 
-        param_options.pop("result")
-        result_widget = True
 
-    def inner_func(func: Callable) -> FunctionGui:
-        from magicgui.widgets import FunctionGui, MainFunctionGui
+class MagicFactory(partial, Generic[_T]):
+    """Factory function that returns a FunctionGui instance.
 
-        cls = MainFunctionGui if main_window else FunctionGui
+    While this can be used directly, (see example below) the preferred usage is
+    via the :func:`magic_factory` decorator.
 
-        func_gui = cls(
-            function=func,
-            call_button=call_button,
-            layout=layout,
-            labels=labels,
-            tooltips=tooltips,
-            param_options=param_options,
-            auto_call=auto_call,
-            result_widget=result_widget,
-            app=app,
-        )
-        func_gui.__wrapped__ = func
-        return func_gui
+    Examples
+    --------
+    >>> def func(x: int, y: str):
+    ...     pass
+    ...
+    >>> factory = MagicFactory(function=func, labels=False)
+    >>> # factory accepts all the same arguments as magicgui()
+    >>> widget1 = factory(call_button=True)
+    >>> # can also override magic_kwargs that were provided when creating the factory
+    >>> widget2 = factory(auto_call=True, labels=True)
+    """
 
-    if function is None:
-        return inner_func
-    else:
-        return inner_func(function)
+    def __new__(cls, function, *args, magic_class=FunctionGui, **keywords):
+        """Create new MagicFactory."""
+        if not function:
+            raise TypeError(
+                "MagicFactory missing required positional argument 'function'"
+            )
+        if isinstance(function, FunctionType):
+            if "<locals>" in function.__qualname__:
+                freevars = function.__code__.co_freevars
+                if function.__name__ in freevars:
+                    warn(
+                        "Self-reference detected in MagicFactory function created "
+                        "in a local scope. FunctionGui references will not work."
+                    )
```

Comment:
This is just so insane. :smile:

Reply:
Indeed, I would suggest adding a comment above the `freevars = ` line to explain wth you are doing here. I would have exactly *zero* idea if I hadn't been the cause of this thing. :joy:

Reply:
![image](https://user-images.githubusercontent.com/1609449/105562468-3a95dc80-5ce8-11eb-9b94-b75b57526492.png)

??